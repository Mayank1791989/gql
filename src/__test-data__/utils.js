/* @flow */
import path from 'path';
import { dedent } from 'dentist';
import splitLines from '../shared/splitLines';

export function getSchema() {
  jest.mock('../shared/watch');
  const watch = require('../shared/watch').default;
  const GQLService = require('../').GQLService;

  const gql = new GQLService({ cwd: path.resolve('src/__test-data__/') });
  // $FlowDisableNextLine
  watch.__triggerChange();
  return gql._schemaBuilder.getSchema();
}

export function getDefLocations() {
  return {
    Player: {
      start: { line: 12, column: 1 },
      end: { line: 16, column: 2 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },
    Player_id: {
      start: { line: 13, column: 3 },
      end: { line: 13, column: 10 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },
    Player_image_arg_size: {
      start: { line: 15, column: 9 },
      end: { line: 15, column: 19 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },

    Team: {
      start: { line: 17, column: 1 },
      end: { line: 20, column: 2 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },
    Node: {
      start: { line: 8, column: 1 },
      end: { line: 10, column: 2 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },
    Edge: {
      start: { line: 27, column: 1 },
      end: { line: 30, column: 2 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },

    Mutation: {
      start: { line: 2, column: 1 },
      end: { line: 5, column: 2 },
      path: path.resolve('src/__test-data__/schema/mutation.gql'),
    },
    Mutation_PlayerCreate: {
      start: { line: 4, column: 3 },
      end: { line: 4, column: 63 },
      path: path.resolve('src/__test-data__/schema/mutation.gql'),
    },
    Mutation_PlayerCreateInput: {
      start: { line: 6, column: 1 },
      end: { line: 9, column: 2 },
      path: path.resolve('src/__test-data__/schema/mutation.gql'),
    },
    Mutation_PlayerCreateInput_id: {
      start: { line: 8, column: 3 },
      end: { line: 8, column: 10 },
      path: path.resolve('src/__test-data__/schema/mutation.gql'),
    },

    Query: {
      start: { line: 2, column: 1 },
      end: { line: 6, column: 2 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },
    Query_viewer: {
      start: { line: 5, column: 3 },
      end: { line: 5, column: 18 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },

    Role: {
      start: { line: 1, column: 1 },
      end: { line: 5, column: 2 },
      path: path.resolve('src/__test-data__/schema/enums.gql'),
    },

    CustomScalar: {
      start: { line: 1, column: 1 },
      end: { line: 1, column: 20 },
      path: path.resolve('src/__test-data__/schema/scalars.gql'),
    },

    customDirective: {
      start: { line: 2, column: 1 },
      end: { line: 7, column: 25 },
      path: path.resolve('src/__test-data__/schema/directives.gql'),
    },
    customDirective_argIf: {
      start: { line: 4, column: 3 },
      end: { line: 4, column: 15 },
      path: path.resolve('src/__test-data__/schema/directives.gql'),
    },
  };
}

export function getRefLocations() {
  return {
    Player: [
      {
        start: { line: 14, column: 11 },
        end: { line: 14, column: 17 },
        path: path.resolve('src/__test-data__/schema/mutation.gql'),
      },
      {
        start: { line: 24, column: 7 },
        end: { line: 24, column: 13 },
        path: path.resolve('src/__test-data__/schema/query.gql'),
      },
      {
        start: { line: 12, column: 6 },
        end: { line: 12, column: 12 },
        path: path.resolve('src/__test-data__/schema/query.gql'),
      },
    ],

    Edge: [
      {
        start: { line: 28, column: 11 },
        end: { line: 28, column: 15 },
        path: path.resolve('src/__test-data__/schema/query.gql'),
      },
    ],

    Role: [
      {
        start: { line: 10, column: 9 },
        end: { line: 10, column: 13 },
        path: path.resolve('src/__test-data__/schema/mutation.gql'),
      },
      {
        start: { line: 34, column: 9 },
        end: { line: 34, column: 13 },
        path: path.resolve('src/__test-data__/schema/query.gql'),
      },
      {
        start: { line: 36, column: 15 },
        end: { line: 36, column: 19 },
        path: path.resolve('src/__test-data__/schema/query.gql'),
      },
      {
        start: { line: 1, column: 6 },
        end: { line: 1, column: 10 },
        path: path.resolve('src/__test-data__/schema/enums.gql'),
      },
    ],

    CustomScalar: [
      {
        start: { line: 35, column: 10 },
        end: { line: 35, column: 22 },
        path: path.resolve('src/__test-data__/schema/query.gql'),
      },
      {
        start: { line: 1, column: 8 },
        end: { line: 1, column: 20 },
        path: path.resolve('src/__test-data__/schema/scalars.gql'),
      },
    ],
  };
}

export function getHints() {
  return {
    ObjectTypes: [
      {
        description: 'Mutation contains all allowed mutations',
        text: 'Mutation',
        type: 'Object',
      },
      {
        description: '',
        text: 'PlayerCreatePayload',
        type: 'Object',
      },
      {
        description: 'Query is the root query object',
        text: 'Query',
        type: 'Object',
      },
      {
        description: '',
        text: 'Player',
        type: 'Object',
      },
      {
        description: '',
        text: 'Team',
        type: 'Object',
      },
      {
        description: '',
        text: 'Viewer',
        type: 'Object',
      },
      {
        description: '',
        text: 'NewPlayer',
        type: 'Object',
      },
    ],

    OutputTypes: [
      {
        description: '',
        text: 'Role',
        type: 'Enum',
      },
      {
        description: 'Mutation contains all allowed mutations',
        text: 'Mutation',
        type: 'Object',
      },
      {
        description: '',
        text: 'PlayerCreatePayload',
        type: 'Object',
      },
      {
        description: 'Query is the root query object',
        text: 'Query',
        type: 'Object',
      },
      {
        description: '',
        text: 'Node',
        type: 'Interface',
      },
      {
        description: '',
        text: 'Player',
        type: 'Object',
      },
      {
        description: '',
        text: 'Team',
        type: 'Object',
      },
      {
        description: '',
        text: 'Viewer',
        type: 'Object',
      },
      {
        description: '',
        text: 'Edge',
        type: 'Interface',
      },
      {
        description: '',
        text: 'NewPlayer',
        type: 'Object',
      },
      {
        description: '',
        text: 'CustomScalar',
        type: 'Scalar',
      },
      {
        description: 'The `String` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.',
        text: 'String',
        type: 'Scalar',
      },
      {
        description: 'The `Int` scalar type represents non-fractional signed whole numeric values. Int can represent values between -(2^31) and 2^31 - 1. ',
        text: 'Int',
        type: 'Scalar',
      },
      {
        description: 'The `Float` scalar type represents signed double-precision fractional values as specified by [IEEE 754](http://en.wikipedia.org/wiki/IEEE_floating_point). ',
        text: 'Float',
        type: 'Scalar',
      },
      {
        description: 'The `Boolean` scalar type represents `true` or `false`.',
        text: 'Boolean',
        type: 'Scalar',
      },
      {
        description: 'The `ID` scalar type represents a unique identifier, often used to refetch an object or as key for a cache. The ID type appears in a JSON response as a String; however, it is not intended to be human-readable. When expected as an input type, any string (such as `"4"`) or integer (such as `4`) input value will be accepted as an ID.',
        text: 'ID',
        type: 'Scalar',
      },
    ],

    CompositeTypes: [
      {
        description: 'Mutation contains all allowed mutations',
        text: 'Mutation',
        type: 'Object',
      },
      {
        description: '',
        text: 'PlayerCreatePayload',
        type: 'Object',
      },
      {
        description: 'Query is the root query object',
        text: 'Query',
        type: 'Object',
      },
      {
        description: '',
        text: 'Node',
        type: 'Interface',
      },
      {
        description: '',
        text: 'Player',
        type: 'Object',
      },
      {
        description: '',
        text: 'Team',
        type: 'Object',
      },
      {
        description: '',
        text: 'Viewer',
        type: 'Object',
      },
      {
        description: '',
        text: 'Edge',
        type: 'Interface',
      },
      {
        description: '',
        text: 'NewPlayer',
        type: 'Object',
      },
    ],

    InputTypes: [
      {
        description: '',
        text: 'Role',
        type: 'Enum',
      },
      {
        description: '',
        text: 'PlayerCreateInput',
        type: 'Input',
      },
      {
        description: '',
        text: 'CustomScalar',
        type: 'Scalar',
      },
      {
        description: 'The `String` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.',
        text: 'String',
        type: 'Scalar',
      },
      {
        description: 'The `Int` scalar type represents non-fractional signed whole numeric values. Int can represent values between -(2^31) and 2^31 - 1. ',
        text: 'Int',
        type: 'Scalar',
      },
      {
        description: 'The `Float` scalar type represents signed double-precision fractional values as specified by [IEEE 754](http://en.wikipedia.org/wiki/IEEE_floating_point). ',
        text: 'Float',
        type: 'Scalar',
      },
      {
        description: 'The `Boolean` scalar type represents `true` or `false`.',
        text: 'Boolean',
        type: 'Scalar',
      },
      {
        description: 'The `ID` scalar type represents a unique identifier, often used to refetch an object or as key for a cache. The ID type appears in a JSON response as a String; however, it is not intended to be human-readable. When expected as an input type, any string (such as `"4"`) or integer (such as `4`) input value will be accepted as an ID.',
        text: 'ID',
        type: 'Scalar',
      },
    ],

    InterfaceTypes: [
      {
        description: '',
        text: 'Node',
        type: 'Interface',
      },
      {
        description: '',
        text: 'Edge',
        type: 'Interface',
      },
    ],

    PlayerTypeFields: [
      {
        description: '',
        text: 'id',
        type: 'ID!',
      },
      {
        description: '',
        text: 'name',
        type: 'String!',
      },
      {
        description: '',
        text: 'image',
        type: 'String!',
      },
    ],

    PlayerTypeImageFieldArgs: [
      {
        description: '',
        text: 'size',
        type: 'Int!',
      },
    ],

    TypesImplementsNode: [
      {
        description: '',
        text: 'Player',
        type: 'Object',
      },
      {
        description: '',
        text: 'Team',
        type: 'Object',
      },
      {
        description: '',
        text: 'Node',
        type: 'Interface',
      },
    ],

    PossibleMutations: [
      {
        description: 'Player Create contains all allowed mutations',
        text: 'PlayerCreate',
        type: 'PlayerCreatePayload',
      },
    ],

    QueryFields: [
      {
        description: '',
        text: 'node',
        type: 'Node',
      },
      {
        description: '',
        text: 'nodes',
        type: '[Node]',
      },
      {
        description: '',
        text: 'viewer',
        type: 'Viewer!',
      },
      {
        description: 'Access the current type schema of this server.',
        text: '__schema',
        type: '__Schema!',
      },
      {
        description: 'Request the type information of a single type.',
        text: '__type',
        type: '__Type',
      },
    ],

    EnumRoleValues: [
      {
        description: '',
        text: 'roleA',
        type: 'Role',
      },
      {
        description: '',
        text: 'roleB',
        type: 'Role',
      },
      {
        description: '',
        text: 'roleC',
        type: 'Role',
      },
    ],
  };
}

export function code(text: string) {
  const sourceText = dedent(text);
  const lines = splitLines(sourceText);
  let position = null;
  lines.forEach((line, index) => {
    const match = line.match(/--\^/);
    if (match) {
      position = {
        // $FlowDisableNextLine
        column: match.index + 3,
        line: index,
      };
    }
  });

  if (!position) {
    throw new Error('Missing --^ in source');
  }

  return {
    sourceText,
    position,
  };
}
