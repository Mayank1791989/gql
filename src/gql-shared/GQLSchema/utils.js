/* @flow */
import {
  type TypeNode,
  type NamedTypeNode,
  type EnumValueDefinitionNode,
  type FieldDefinitionNode,
  getDirectiveValues,
  GraphQLDeprecatedDirective,
  Kind,
} from 'graphql';

export function getNamedTypeNode(typeNode: TypeNode): NamedTypeNode {
  let namedType = typeNode;
  while (
    namedType.kind === Kind.LIST_TYPE ||
    namedType.kind === Kind.NON_NULL_TYPE
  ) {
    namedType = namedType.type;
  }
  return namedType;
}

/**
 * Given a field or enum value node, returns the string value for the
 * deprecation reason.
 */
export function getDeprecationReason(
  node: EnumValueDefinitionNode | FieldDefinitionNode,
): ?string {
  const deprecated = getDirectiveValues(GraphQLDeprecatedDirective, node);
  return deprecated && (deprecated.reason: any);
}
