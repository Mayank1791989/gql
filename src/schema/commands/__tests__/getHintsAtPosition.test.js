/* @flow */
import { getHintsAtPosition } from '../index';
import { getSchema, getHints, code } from '../../../__test-data__/utils';

const schema = getSchema();
const allHints = getHints();
const getHintsAt = (text) => {
  const { sourceText, position } = code(text);
  return getHintsAtPosition(schema, sourceText, position);
};

describe('autocomplete for `type`', () => {
  it('show all types for Field', () => {
    const hints = getHintsAt(`
      type TestType {
        name:   #
        #-----^
      }
    `);
    expect(hints).toEqual(allHints.OutputTypes);
  });

  it('show all types for Field even after typing few chars', () => {
    const hints = getHintsAt(`
      type TestType {
        name: S  #
        #------^
      }
    `);
    expect(hints).toEqual(allHints.OutputTypes);
  });

  it('show all output types for Field if cursor on ! and of the form T!', () => {
    const hints = getHintsAt(`
      type TestType {
        name: S!   ##
        #-------^
      }
    `);
    expect(hints).toEqual(allHints.OutputTypes);
  });

  it('show all output types for Field if cursor on ] and of type [T]', () => {
    const hints = getHintsAt(`
      type TestType {
        name: [S] #
        #-------^
      }
    `);
    expect(hints).toEqual(allHints.OutputTypes);
  });

  it('show all output types for Field if cursor on ! and of the form [T!]', () => {
    const hints = getHintsAt(`
      type TestType {
        name: [S!]
        #-------^
      }
    `);
    expect(hints).toEqual(allHints.OutputTypes);
  });

  it('dont show any types for Field if cursor after !', () => {
    const hints = getHintsAt(`
      type TestType {
        name: S!  #
        #--------^
      }
    `);
    expect(hints).toEqual([]);
  });

  it('show all types for Field with args', () => {
    const hints = getHintsAt(`
      type TestType {
        players(first: String!):   #
                        #--------^
      }
    `);
    expect(hints).toEqual(allHints.OutputTypes);
  });

  it('show all types for Field ListType', () => {
    const hints = getHintsAt(`
      type TestType {
        name: []
        #-----^ (inside bracket)
      }
    `);
    expect(hints).toEqual(allHints.OutputTypes);
  });

  it('show only InputType for Field input args ', () => {
    const hints = getHintsAt(`
      type TestType {
        name(xyz: )
        #--------^
      }
    `);
    expect(hints).toEqual(allHints.InputTypes);
  });

  it('show `implements` && `{` after typename', () => {
    const hints = getHintsAt(`
      type TestType   #
      #-------------^
    `);

    expect(hints).toMatchSnapshot();
  });

  it('show all `GraphQLInterfaceType` types after `implements` keyword', () => {
    const hints = getHintsAt(`
      type TestType implements   # (Dont remote hash will avoid removing space after line)
      #------------------------^
    `);
    expect(hints).toEqual(allHints.InterfaceTypes);
  });

  it('show all `GraphQLInterfaceType` types after multiple InterfaceTypes', () => {
    const hints = getHintsAt(`
      type TestType implements Node   #
            #-----------------------^
    `);
    expect(hints).toEqual(allHints.InterfaceTypes);
  });
});

describe('autocomplete for `input`', () => {
  it('show only InputType for field', () => {
    const hints = getHintsAt(`
      input TestTypeInput {
        name:   #
        ------^
      }
    `);
    expect(hints).toEqual(allHints.InputTypes);
  });

  it('show only InputType for field of ListType', () => {
    const hints = getHintsAt(`
      input TestType {
        name: []
        ------^ (inside bracket)
      }
    `);
    expect(hints).toEqual(allHints.InputTypes);
  });
});

describe('autocomplete for `union`', () => {
  it('show only GrahQLObjectType for field', () => {
    const hints = getHintsAt(`
      union TestUnionType =   #
          #-----------------^
    `);
    expect(hints).toEqual(allHints.ObjectTypes);
  });

  it('show `|` after Type(unionMember)', () => {
    const hints = getHintsAt(`
      union TestUnionType =  TestType   #
                  #-------------------^
    `);
    expect(hints).toMatchSnapshot();
  });

  it('show all GrahQLObjectType after `|`', () => {
    const hints = getHintsAt(`
      union TestUnionType =  TestType |   #
                    #-------------------^
    `);
    expect(hints).toEqual(allHints.ObjectTypes);
  });
});
