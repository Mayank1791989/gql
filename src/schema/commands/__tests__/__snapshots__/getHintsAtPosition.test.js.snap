exports[`autocomplete for \`type\` show \`implements\` && \`{\` after typename 1`] = `
Array [
  Object {
    "text": "implements",
    "type": "Implements",
  },
  Object {
    "text": "{",
  },
]
`;

exports[`autocomplete for \`union\` show \`|\` after Type(unionMember) 1`] = `
Array [
  Object {
    "description": "OR symbol",
    "text": "|",
    "type": "OR",
  },
]
`;
